### DISCUSSION TITLE
Predict a known sinusoidal curve using PySINDY

### DISCUSSION THREAD

**Discussion by azanparmar** [dynamicslab/pysindy#260] [Q&A]
Hello team,

I am trying to predict a sinusoidal curve of 5 Hz with and amplitude of 2 using PySINDY. I have tried multiple libraries and functions but failed to come at a perfect model fit.
Can you please guide me on how it can be done the best?

[SINDY_SineWave.zip](https://github.com/dynamicslab/pysindy/files/9926236/SINDY_SineWave.zip)

Thanks in advance for your help!

**Comment by Jacob-Stevens-Haas** [comment]
Hey @azanparmar , sorry for the slow reply.  Since a sine obeys the differential equation:

$$
\ddot x + \lambda^2 x = 0
$$

You'll want the variable `x` to represent a 3-vector of [position, velocity, acceleration], so your data matrix should be shape (n, 3), where n is the number of points.  The standard function library (`PolynomialLibrary`) should work.

If you can't "know" it's a 2nd order ODE prior, you could try (a) letting `x` represent a vector of time-lags, or (b) setting time as a control input and using the `FourierLibrary`.

**Reply by azanparmar** [reply]
Hello @Jacob-Stevens-Haas, thank you for your reply.

I tried what you mentioned in the reply. Please refer the image:

![image](https://user-images.githubusercontent.com/117339917/202706912-bb9026d2-180a-471b-a0d1-3e0e35085103.png)


The model predicts:
![image](https://user-images.githubusercontent.com/117339917/202706971-8a83eae5-3a2c-48ea-9aba-aee7d690fa29.png)

But this is obvious, isn't it?

It is fitting your x' in x_dot and x'' in x_double_dot data. 

Is this what we should be expecting?

**Reply by azanparmar** [reply]
@Jacob-Stevens-Haas Hello Sir,

I guess I am getting what you are trying to say. 

For x''+ (lambda^2)*x = 0, we must just specify [position,velocity] = [x,x'] because SINDY will predict x' and (x')' in terms of x and x' which will eventually solve the problem which we are trying to solve.

If we provide acceleration x'', then sindy will predict x''' in terms of x,x',x'' which makes life even more complicated.

I tried just with position and velocity and it led me to
![image](https://user-images.githubusercontent.com/117339917/202724334-fed0ea22-89d7-47ed-a05d-356326f08e7c.png)

This ode x'' = -986.96*x leads to a solution of rotational frequency 31.41 rad/s which is 4.999 Hz which is closest to the input frequency 5 Hz.

I am facing a strange issue now, how to I go back from x'' to x.

I can use model.simulate to obtain x? 

Can you help me with that?

**Reply by Jacob-Stevens-Haas** [reply]
Great catch - just specifying [x, x'] :smile: 

Yes, `model.simulate()` helps you predict `x`.  Since your input vector is [x, x'], `model.simulate` would give you both of those values, and you would just need to choose the first column.

**Reply by azanparmar** [reply]
Thanks a ton @Jacob-Stevens-Haas !

You helped me solve a problem which I have been trying since a very long time! Kudos!
